external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala

```diff
@@ -18,34 +18,58 @@
 package org.apache.spark.sql.kafka010
 
 import java.{util => ju}
+import java.util.concurrent.{ConcurrentMap, TimeUnit}
 import javax.annotation.concurrent.GuardedBy
 
+import com.google.common.cache._
+import org.apache.kafka.clients.producer.KafkaProducer
+import scala.collection.JavaConverters._
 import scala.collection.immutable.SortedMap
 import scala.collection.mutable
-
-import org.apache.kafka.clients.producer.KafkaProducer
+import scala.util.control.NonFatal
 
 import org.apache.spark.internal.Logging
 
 private[kafka010] object CachedKafkaProducer extends Logging {
 
   private type Producer = KafkaProducer[Array[Byte], Array[Byte]]
 
-  @GuardedBy("this")
-  private val cacheMap = new mutable.HashMap[String, Producer]()
+  private val cacheExpireTimeout: Long = System.getProperty("spark.kafka.guava.cache.timeout",
+    "10").toLong
+
+  private val noneReturningLoader = new CacheLoader[String, Option[Producer]] {
```

Comment:
The builder shouldn't **require** a loader.
https://google.github.io/guava/releases/16.0/api/docs/com/google/common/cache/CacheBuilder.html#build()
Not sure we use the same guava version though